In [ ]:
import pandas as pd

df = pd.read_csv("/content/Student_Performance.csv")

print("Kích thước ban đầu:", df.shape)
print("Số dòng trùng:", df.duplicated().sum())

Kích thước ban đầu: (10000, 6)
Số dòng trùng: 127


In [ ]:
df = df.drop_duplicates().reset_index(drop=True)

print("Kích thước sau khi xóa duplicate:", df.shape)
print("Số dòng trùng còn lại:", df.duplicated().sum())

Kích thước sau khi xóa duplicate: (9873, 6)
Số dòng trùng còn lại: 0


In [ ]:
print("Giá trị thiếu:")
print(df.isnull().sum())

Giá trị thiếu:
Hours Studied                       0
Previous Scores                     0
Extracurricular Activities          0
Sleep Hours                         0
Sample Question Papers Practiced    0
Performance Index                   0
dtype: int64


In [ ]:
X = df.drop(columns=["Performance Index"])
y = df["Performance Index"]

print("X shape:", X.shape)
print("y shape:", y.shape)
print("\nCác biến đầu vào:")
print(X.columns.tolist())

X shape: (9873, 5)
y shape: (9873,)

Các biến đầu vào:
['Hours Studied', 'Previous Scores', 'Extracurricular Activities', 'Sleep Hours', 'Sample Question Papers Practiced']


In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print("X_train:", X_train.shape)
print("X_test:", X_test.shape)
print("y_train:", y_train.shape)
print("y_test:", y_test.shape)

X_train: (7898, 5)
X_test: (1975, 5)
y_train: (7898,)
y_test: (1975,)


In [ ]:
numeric_features = [
    "Hours Studied",
    "Previous Scores",
    "Sleep Hours",
    "Sample Question Papers Practiced"
]

categorical_features = [
    "Extracurricular Activities"
]

print("Numeric features:", numeric_features)
print("Categorical features:", categorical_features)

Numeric features: ['Hours Studied', 'Previous Scores', 'Sleep Hours', 'Sample Question Papers Practiced']
Categorical features: ['Extracurricular Activities']


In [ ]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder

preprocessor = ColumnTransformer(
    transformers=[
        ("num", StandardScaler(), numeric_features),
        ("cat", OneHotEncoder(handle_unknown="ignore"), categorical_features)
    ]
)

print("Preprocessing pipeline đã được tạo.")

Preprocessing pipeline đã được tạo.


In [ ]:
X_train_processed = preprocessor.fit_transform(X_train)
X_test_processed = preprocessor.transform(X_test)

print("X_train sau preprocessing:", X_train_processed.shape)
print("X_test sau preprocessing:", X_test_processed.shape)

X_train sau preprocessing: (7898, 6)
X_test sau preprocessing: (1975, 6)


In [ ]:
feature_names = preprocessor.get_feature_names_out()

print("Các feature sau preprocessing:")
for i, name in enumerate(feature_names, start=1):
    print(f"{i}. {name}")

Các feature sau preprocessing:
1. num__Hours Studied
2. num__Previous Scores
3. num__Sleep Hours
4. num__Sample Question Papers Practiced
5. cat__Extracurricular Activities_No
6. cat__Extracurricular Activities_Yes


In [ ]:
print("Kiểm tra preprocessing:")
print("X_train_processed shape:", X_train_processed.shape)
print("X_test_processed shape:", X_test_processed.shape)
print("Số lượng feature:", len(feature_names))

Kiểm tra preprocessing:
X_train_processed shape: (7898, 6)
X_test_processed shape: (1975, 6)
Số lượng feature: 6


## Kết luận preprocessing

- Dataset ban đầu có 10.000 dòng.
- Phát hiện 127 dòng trùng lặp và đã loại bỏ, còn 9.873 dòng.
- Không có giá trị thiếu nên không cần thực hiện imputation.
- Biến mục tiêu `Performance Index` được tách riêng khỏi các biến đầu vào.
- Dữ liệu được chia thành 80% tập train và 20% tập test.
- 4 biến số được chuẩn hóa bằng `StandardScaler`.
- Biến phân loại `Extracurricular Activities` được mã hóa bằng `OneHotEncoder`.
- Sau preprocessing, dữ liệu có 6 features.
- Bộ preprocessing chỉ được `fit` trên tập train và chỉ `transform` trên tập test để tránh data leakage.
- Các bước preprocessing sẽ được tích hợp vào pipeline khi huấn luyện mô hình để đảm bảo dữ liệu đầu vào khi triển khai được xử lý nhất quán.